### ISSUE TITLE
python@2 opt link lost in migration (python3 installation removes it)

### ISSUE THREAD

**Issue by philpennock** [Homebrew/homebrew-core#24978]
- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? - [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

I ran `brew upgrade`, which updated `mercurial`; that then failed, with:
```
Error: No such file or directory - /usr/local/opt/mercurial/bin/hg
```

Uninstall, reinstall, persist; file exists though.  It's an interpreter error:

```
% hg
zsh: /usr/local/bin/hg: bad interpreter: /usr/local/opt/python@2/bin/python2.7: no such file or directory
```

It _appears_ that the "python is python3, use python@2 for 2.7" migration did not link `python@2` and `brew doctor` does not fix it.

```
% ls -ld /usr/local/opt/py*
lrwxr-xr-x  1 pdp  admin  21 Mar  6 11:34 /usr/local/opt/pyenv -> ../Cellar/pyenv/1.2.2
lrwxr-xr-x  1 pdp  admin  32 Oct 12 09:36 /usr/local/opt/pyenv-virtualenv -> ../Cellar/pyenv-virtualenv/1.1.1
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python3 -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python@3 -> ../Cellar/python/3.6.4_3
```

Running: `brew reinstall python@2` **does fix it**.   Afterwards:

```
% ls -ld /usr/local/opt/py*
lrwxr-xr-x  1 pdp  admin  21 Mar  6 11:34 /usr/local/opt/pyenv -> ../Cellar/pyenv/1.2.2
lrwxr-xr-x  1 pdp  admin  32 Oct 12 09:36 /usr/local/opt/pyenv-virtualenv -> ../Cellar/pyenv-virtualenv/1.1.1
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 pdp  admin  27 Mar  7 18:31 /usr/local/opt/python2 -> ../Cellar/python@2/2.7.14_1
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python3 -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 pdp  admin  27 Mar  7 18:31 /usr/local/opt/python@2 -> ../Cellar/python@2/2.7.14_1
lrwxr-xr-x  1 pdp  admin  24 Mar  6 11:34 /usr/local/opt/python@3 -> ../Cellar/python/3.6.4_3
```

So everything is working now, but it looks like there's an opportunity for more automated repair of the migration there.

**Comment by ilovezfs** [comment]
Yes it looks like there's a bug here.
```
iMac-TMP:homebrew-core joe$ git reset --hard 3cbcbf1bf99a8c2f766e3b0196e7937922ed9ce7
HEAD is now at 3cbcbf1 caffe: add 1.0 bottle.
iMac-TMP:homebrew-core joe$ brew install python
==> Downloading https://homebrew.bintray.com/bottles/python-2.7.14_3.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/python-2.7.14_3.el_capitan.bottle.tar.gz
==> Pouring python-2.7.14_3.el_capitan.bottle.tar.gz
==> /usr/local/Cellar/python/2.7.14_3/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/python/
==> /usr/local/Cellar/python/2.7.14_3/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/python/
==> /usr/local/Cellar/python/2.7.14_3/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/python/
==> Caveats
This formula installs a python2 executable to /usr/local/bin.
If you wish to have this formula's python executable in your PATH then add
the following to ~/.bash_profile:
  export PATH="/usr/local/opt/python/libexec/bin:$PATH"

Pip and setuptools have been installed. To update them
  pip2 install --upgrade pip setuptools

You can install Python packages with
  pip2 install <package>

They will install into the site-package directory
  /usr/local/lib/python2.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
==> Summary
🍺  /usr/local/Cellar/python/2.7.14_3: 3,517 files, 48.5MB
iMac-TMP:homebrew-core joe$ brew update &>/dev/null
iMac-TMP:homebrew-core joe$ ls -l /usr/local/opt/|grep py
lrwxr-xr-x  1 joe  admin  25 Mar  7 16:14 python -> ../Cellar/python/2.7.14_3
lrwxr-xr-x  1 joe  admin  25 Mar  7 16:14 python2 -> ../Cellar/python/2.7.14_3
lrwxr-xr-x  1 joe  admin  25 Mar  7 16:14 python@2 -> ../Cellar/python/2.7.14_3
iMac-TMP:homebrew-core joe$ brew install python@2
==> Downloading https://homebrew.bintray.com/bottles/python@2-2.7.14_1.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/python@2-2.7.14_1.el_capitan.bottle.tar.gz
==> Pouring python@2-2.7.14_1.el_capitan.bottle.tar.gz
==> /usr/local/Cellar/python@2/2.7.14_1/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/pytho
==> /usr/local/Cellar/python@2/2.7.14_1/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/pytho
==> /usr/local/Cellar/python@2/2.7.14_1/bin/python2 -s setup.py --no-user-cfg install --force --verbose --single-version-externally-managed --record=installed.txt --install-scripts=/usr/local/Cellar/pytho
==> Caveats
This formula installs a python2 executable to /usr/local/opt/python@2/bin
If you wish to have this formula's python executable in your PATH then add
the following to ~/.bash_profile:
  export PATH="/usr/local/opt/python@2/libexec/bin:$PATH"

Pip and setuptools have been installed. To update them
  pip2 install --upgrade pip setuptools

You can install Python packages with
  pip2 install <package>

They will install into the site-package directory
  /usr/local/lib/python2.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python

This formula is keg-only, which means it was not symlinked into /usr/local,
because this is an alternate version of another formula.

If you need to have this software first in your PATH run:
  echo 'export PATH="/usr/local/opt/python@2/bin:$PATH"' >> ~/.bash_profile

==> Summary
🍺  /usr/local/Cellar/python@2/2.7.14_1: 3,523 files, 48.6MB
iMac-TMP:homebrew-core joe$ ls -l /usr/local/opt/|grep py
lrwxr-xr-x  1 joe  admin  25 Mar  7 16:14 python -> ../Cellar/python/2.7.14_3
lrwxr-xr-x  1 joe  admin  27 Mar  7 16:15 python2 -> ../Cellar/python@2/2.7.14_1
lrwxr-xr-x  1 joe  admin  27 Mar  7 16:15 python@2 -> ../Cellar/python@2/2.7.14_1
iMac-TMP:homebrew-core joe$ brew upgrade python
==> Upgrading 1 outdated package, with result:
python 3.6.4_3
==> Upgrading python 
==> Installing dependencies for python: xz
==> Installing python dependency: xz
==> Downloading https://homebrew.bintray.com/bottles/xz-5.2.3.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/xz-5.2.3.el_capitan.bottle.tar.gz
==> Pouring xz-5.2.3.el_capitan.bottle.tar.gz
🍺  /usr/local/Cellar/xz/5.2.3: 92 files, 1.4MB
==> Installing python
==> Downloading https://homebrew.bintray.com/bottles/python-3.6.4_3.el_capitan.bottle.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/python-3.6.4_3.el_capitan.bottle.tar.gz
==> Pouring python-3.6.4_3.el_capitan.bottle.tar.gz
==> /usr/local/Cellar/python/3.6.4_3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.6.4_3/bin --install-lib=/usr/local/lib/python3.6/site-pack
==> /usr/local/Cellar/python/3.6.4_3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.6.4_3/bin --install-lib=/usr/local/lib/python3.6/site-pack
==> /usr/local/Cellar/python/3.6.4_3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.6.4_3/bin --install-lib=/usr/local/lib/python3.6/site-pack
==> Caveats
Pip, setuptools, and wheel have been installed. To update them
  pip3 install --upgrade pip setuptools wheel

You can install Python packages with
  pip3 install <package>

They will install into the site-package directory
  /usr/local/lib/python3.6/site-packages

See: https://docs.brew.sh/Homebrew-and-Python

Unversioned symlinks python, python-config, pip etc. pointing to python3,
python3-config, pip3 etc., respectively, have been installed.

If you need Homebrew's Python 2, `brew install python@2`.
==> Summary
🍺  /usr/local/Cellar/python/3.6.4_3: 3,615 files, 56.3MB
iMac-TMP:homebrew-core joe$ ls -l /usr/local/opt/|grep py
lrwxr-xr-x  1 joe  admin  24 Mar  7 16:15 python -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 joe  admin  24 Mar  7 16:15 python3 -> ../Cellar/python/3.6.4_3
lrwxr-xr-x  1 joe  admin  24 Mar  7 16:15 python@3 -> ../Cellar/python/3.6.4_3
iMac-TMP:homebrew-core joe$ 
```

So the opt link for python@2 is created but is then deleted when python gets upgraded. CC @MikeMcQuaid

**Comment by ideologysec** [comment]
`brew uninstall --force python2; brew install python2`

is also broken for me - I have no python2 executable in /usr/local/bin which is driving me up a wall (I don't want to have to put the keg path in my $PATH; that sets "python --version" to python2)

**Comment by ilovezfs** [comment]
@ideologysec define not working.

**Comment by ilovezfs** [comment]
The specific issue is whether /usr/local/opt/python@2 exists. I assume it does, and you're just remarking on the fact that python@2 is keg_only now, which is irrelevant to what this issue is about.

**Comment by ideologysec** [comment]
You're right. I misread the warning at the end, which changed at some point to "/usr/local/opt/python@2/bin/python" instead of "/usr/local/bin/python2." Appears that I could sort this out via a symlink myself.

I did have an issue with the (forced) upgrade to the new python schema about a week ago, since it broke my python setup, but I'm unsure what logs I could provide at this point that would be helpful, so I'll apologize and bow out of the discussion.

**Comment by ilovezfs** [comment]
@ideologysec no worries. I suggest you just `brew link -f python@2` if you want the links in /usr/local/bin.

**Comment by ilovezfs** [comment]
@philpennock this should fix it: https://github.com/Homebrew/brew/pull/3893

**Comment by RandomDSdevel** [comment]
@ilovezfs:  
> Are you sure? The uninstall + install should work.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Never mind, that was just me being silly and losing my marbles for the umpteenth time.  Apparently I was messing around with the _normal_ `python` formula _instead_ of the `python@2` one to see whether the former's `python3` and `python@3` symbolic links in its `opt_prefix` to its own folder there were relics left over from how things were set up before the '`python3` → `python`' migration, when `python3` was a formula and `python@3` was an alias for it, but both adjusted to point to `python` _after_ the migration versus the new status quo.  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;So, an `uninstall` and `install` did work to fix this issue as it manifested itself on my system.  (I tried a `reinstall` first, but then `brew` complained about some files from the old `python@2` still being present during the `--build-from-source` and, after raising several exceptions which I just ignored, finally balked and bailed out on said build.)  What's funny about what I saw before that, though, is this:  I installed `python@2` after running `brew update` after the migration was primed in the source copy of this repository hosted upstream here on GitHub and running `brew cleanup`…and then `brew`, strangely enough, didn't properly link `python@2` into its `opt_prefix`.
